contrib/profiler/profiler.py

```diff
@@ -188,148 +143,188 @@ def get_network_bytes(filename):
-            max_mem_idx = i
-        if sample_list[i].get_read_bytes() > max_read:
-            max_read = sample_list[i].get_read_bytes()
-            max_read_idx = i
-
-        if sample_list[i].get_gpu_usage() <= 10:
-            cpu_when_gpu_low.append(sample_list[i].get_cpu_usage())
-            mem_when_gpu_low.append(sample_list[i].get_mem_used())
-            disk_read_when_gpu_low.append(sample_list[i].get_read_bytes())
-
-        sum_cpu += sample_list[i].get_cpu_usage()
-        sum_gpu += sample_list[i].get_gpu_usage()
-        sum_mem += sample_list[i].get_mem_used()
-        sum_read += sample_list[i].get_read_bytes()
-
-    length_gpu_low = len(cpu_when_gpu_low)
-    if length_gpu_low == 0:
-        length_gpu_low = 1
-
-    print('%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f' %
-          (max_gpu, sum_gpu / count, max_cpu, sum_cpu / count, max_mem, sum_mem / count, max_read, sum_read,
-           sum(cpu_when_gpu_low) / length_gpu_low,
-           sum(mem_when_gpu_low) / length_gpu_low,
-           sum(disk_read_when_gpu_low) / length_gpu_low))
-    return [max_gpu, sum_gpu / count, max_cpu, sum_cpu / count, max_mem, sum_mem / count, max_read, sum_read,
-            sum(cpu_when_gpu_low) / length_gpu_low,
-            sum(mem_when_gpu_low) / length_gpu_low,
-            sum(disk_read_when_gpu_low) / length_gpu_low]
+def analyze_samples(sample_list, advise):
+    sample_list = np.array(sample_list, dtype=np.float)
+    used_gpu_num = (sample_list.shape[1] - 7) / 4
+    cpu_usage = sample_list[:, 0]
+    mem_usage = sample_list[:, 1] / sample_list[:, 2]
+    gpu_usage = list()
+    gpu_mem_usage = list()
+    for i in range(int(used_gpu_num)):
+        gpu_usage.append(sample_list[:, 7 + i * 4])
+        gpu_mem_usage.append(sample_list[:, 9 + i * 4] / sample_list[:, 10 + i * 4])
+
+    if used_gpu_num >= 2:
+        # multiple GPUs, analyze whether each GPU has the same memory usage
+        gpu_mem_avg_0 = np.average(gpu_mem_usage[0])
+        gpu_mem_avg_1 = np.average(gpu_mem_usage[1])
+        if abs(gpu_mem_avg_0 - gpu_mem_avg_1) > 0.15:
+            advise.add_times(index=1)
+
+    if np.average(gpu_usage[0]) >= 90:
+        advise.add_times(index=0)
+    else:
+        if np.average(gpu_mem_usage[0]) < 0.80:
+            advise.add_times(index=2)
+        else:
+            advise.add_times(index=3)
+        # slide the cpu and get the value to divide the cpu value
+        slide_windows = SlideWindows(10)
+        cpu_slide = list()
+        for i in range(cpu_usage.shape[0]):
+            cpu_slide.append(slide_windows.get_data(cpu_usage[i]))
+        cpu_slide_copy = cpu_slide.copy()
+        cpu_slide_copy.sort()
+        cpu_std_max = cpu_slide_copy[int(len(cpu_slide_copy) * 0.8)]
+        cpu_std_min = cpu_slide_copy[int(len(cpu_slide_copy) * 0.2)]
+
+        gpu_usage_up_down = [0]
+        for i in range(1, len(gpu_usage[0])):
+            if gpu_usage[0][i] > gpu_usage[0][i - 1]:
```

Comment:
Do we need to add a threshold to descide up or down? If the change is minor, shall we ignore it?

Reply:
According to the testing data, the change in the values of GPU usage is either zero or not minor. I think we don't need to set a threshold.

Reply:
OK, fine

---
In reply to: [332372450](https://github.com/microsoft/pai/pull/3649#discussion_r332372450) [](ancestors = 332372450)